quinn/src/endpoint.rs

```diff
@@ -490,25 +504,59 @@ impl State {
                     }
                 }
                 Poll::Pending => {
-                    break;
+                    return (received_data, Ok(false));
                 }
                 // Ignore ECONNRESET as it's undefined in QUIC and may be injected by an
                 // attacker
                 Poll::Ready(Err(ref e)) if e.kind() == io::ErrorKind::ConnectionReset => {
                     continue;
                 }
                 Poll::Ready(Err(e)) => {
-                    return Err(e);
+                    return (received_data, Err(e));
                 }
             }
             if !self.recv_limiter.allow_work() {
-                self.recv_limiter.finish_cycle();
-                return Ok(true);
+                return (received_data, Ok(true));
             }
         }
+    }
 
-        self.recv_limiter.finish_cycle();
-        Ok(false)
+    fn drive_recv<'a>(&'a mut self, cx: &mut Context, now: Instant) -> Result<bool, io::Error> {
```

Comment:
We generally prefer to order functions top-down, so `drive_recv()` should come before `poll_socket()`.